BJS2-16317

fraifree opened the pull request:
create SubscriptionController
with methods:
followUser();
unFollowUser();
getFollowers();
getFollowing();
getFollowersCount();
getFollowingCount();
create UserFilter interface with filters;
refactor SubscriptionRepository;
add SubscriptionControllerTest and SubscriptionServiceTest

fraifree committed:
create SubscriptionService
create SubscriptionController
with methods:
followUser();
unFollowUser();
getFollowers();
getFollowing();
getFollowersCount();
getFollowingCount();
create UserFilter interface with filters;
refactor SubscriptionRepository;
add SubscriptionControllerTest and SubscriptionServiceTest

fokinms commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    @Transactional
```
@Transactional здесь на мой взгляд не нужен. метод не вносит никаких изменений в БД

InnMish replied:
Если уже используешь транзакции, используем как следует 
Если метод только на чтение, устанавливаем параметр readOnly

fokinms commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+        return getUserDtos(filters, followers);
+    }
+
+    @Transactional
```
the same... @transactional

InnMish replied:
readOnly

Commit:
fixed gradle file

Commit:
Merge branch 'griffon-master-bc5' into griffon-master-bc5-BJS2-16317

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+public class SubscriptionController {
+    private final SubscriptionService subscriptionService;
+
+
```
вот тут отступ после класса, затем между полем и методом отступ в одну строчку

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+import java.util.Objects;
+
+@RestController
+@RequestMapping(value = "/subscripts")
```
subsription  

слово subscripts не отражает того, что тут происходит, потому что script - другое слово совсем)

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+
+
+    @RequestMapping(value = "/followUser",
+            method = RequestMethod.POST,
```
вешай аннотацию PostMapping, тогда не придется параметр method  тут прописывать у аннотации @RequestMapping

По пути: по REST не принято писать в camelCase, поэтому если хочешь в маппинге писать именно followUser, пиши через слэш - follow/user

Далее: ты сейчас айдишники передаешь в формате RequestParam, но я бы рекомендовала прокидывать их через @PathValiable, чтобы мы в пути сразу увидели, кто и на кого пытается подписаться. 

Это тема следующей недели, но раз ты уже это пишешь, давай делать)

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+
+    @RequestMapping(value = "/followUser",
+            method = RequestMethod.POST,
+            produces = MediaType.APPLICATION_JSON_VALUE)
```
это стандартный тип ответа, можно не прописывать, если это и так по умолчанию спрингом возвращается

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+    @RequestMapping(value = "/followUser",
+            method = RequestMethod.POST,
+            produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<String> followUser(@RequestParam(value = "followerId") Long followerId,
```
Насчет ResponseEntity<String> 

Лично я всегда возвращаю просто dto или void, если по ресту метод не подразумевает возвращаемого значения. 

Если здесь выбираем делать через ResponseEntity<String>, нужно договориться с командой о едином код-стайле в User_service. На данном этапе это может быть немного проблематичо, потому что все ребята на разном уровне, и придется объяснять принцип работы ResponseEntity

Предлагаю в этом сервисе не запариваться с форматом ответа, ва возвращать dto или void 

А вот дальше, начиная где-то с post_service (след. неделя) можем поговорить об использовании ResponseEntity

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+
+    @RequestMapping(value = "/unFollowUser",
+            method = RequestMethod.POST,
+            produces = MediaType.APPLICATION_JSON_VALUE)
```
unfollow - одно слово 

unfollow/{followeeId}/{followerId} - примерно к такому формату нужно прийти 

метод не прописываем, используем постмаппинг, produces и так по умолчанию

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+    @RequestMapping(value = "/getFollowers",
+            method = RequestMethod.GET,
+            produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<List<UserDto>> getFollowers(@RequestParam(value = "followerId") Long followerId,
```
через @PathValiable

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    @RequestMapping(value = "/getFollowers",
```
/get/{userId}/followers 

получить/чьих/подписчиков

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+
+    @RequestMapping(value = "/getFollowers",
+            method = RequestMethod.GET,
+            produces = MediaType.APPLICATION_JSON_VALUE)
```
гетмаппинг, без produces

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+    @RequestMapping(value = "/getFollowing",
+            method = RequestMethod.GET,
+            produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<List<UserDto>> getFollowing(@RequestParam(value = "followeeId") Long followeeId,
```
тут также, как выше в комментах

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+    @RequestMapping(value = "/getFollowersCount",
+            method = RequestMethod.GET,
+            produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<Long> getFollowersCount(@RequestParam(value = "followerId") Long followerId) {
```
также, как выше

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,85 @@
+    @RequestMapping(value = "/getFollowingCount",
+            method = RequestMethod.GET,
+            produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<Long> getFollowingCount(@RequestParam(value = "followerId") Long followerId) {
```
также, как выше

InnMish commented on src/main/java/school/faang/user_service/filter/UserFilter.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.List;
+import java.util.stream.Stream;
+
```
строчку лишнюю убираем) 
это мелочь, но я все равно пишу) код должен быть красивым во всем хехе

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+    @Transactional
+    public void followUser(long followerId, long followeeId) {
+        if (subscriptionRepository.existsByFollowerIdAndFolloweeId(followerId, followeeId)) {
+            log.error("Method followUser get exception");
```
хехе 
смотри, логи - это для разработчиков приложения 

здесь важно строго разделять, что для разработчика ошибка, что предупреждение, а что просто лог для информации 

если пользователь закинул какие-то невалидные данные, например, как тут - пытается подписаться на того, на кого и так подписан, разве для разработчичка (для нас) это проблема? Нет, это не ошибка, это пользователь просто ерундой занимается 

А мы стригеримся на этого лог еррор и побежим что-то чинить в своем приложении. А там... ничего не сломалось) 

Поэтому здесь должен быть log.info

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Transactional
+    public void unFollowUser(long followerId, long followeeId) {
```
unfollowUser

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+        } else {
+            log.error("Method unFollowUser get exception");
+            throw new DataValidationException("Subscription not found");
+        }
```
а почему в разном код-стайле пишешь?) У тебя выше в методе такая же логика, за исключением проверки на false  

выбери какой-то один вариант) я за этот. Сначала в методе видно основную логику, а потом уже все, что делать, если поломка какая-то

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+        if (subscriptionRepository.existsByFollowerIdAndFolloweeId(followerId, followeeId)) {
+            subscriptionRepository.unfollowUser(followerId, followeeId);
+        } else {
+            log.error("Method unFollowUser get exception");
```
тоже тут log.info

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+    @Transactional
+    public List<UserDto> getFollowers(long followerId, UserFilterDto filters) {
+        Stream<User> followers = subscriptionRepository.findByFollowerId(followerId);
+        return getUserDtos(filters, followers);
```
Вот если просто читать код здесь, то я читаю так: 
идем в базу, получаем всех подписчиков, и возвращаем список дто эти подписчиков. Зачем-то еще в формирование дто фильтры передаем. 

Так нельзя) название метод getUserDtos вводит в заблуждение 
здесь я бы метод назвала filterUsers

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+    @Transactional
+    public List<UserDto> getFollowing(long followeeId, UserFilterDto filters) {
+        Stream<User> followee = subscriptionRepository.findByFolloweeId(followeeId);
+        return getUserDtos(filters, followee);
```
такой же коммент по неймингу, как выше

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @NotNull
+    private List<UserDto> getUserDtos(UserFilterDto filters, Stream<User> followee) {
```
Stream<User> followee 
наверное, followees?

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+
+    @NotNull
+    private List<UserDto> getUserDtos(UserFilterDto filters, Stream<User> followee) {
+        List<UserFilter> userFiltersList = userFilters.stream()
```
я по типу этого поля уже виду, что это лист, в названии можно не писать еще раз List 
Поле я бы назвала не просто какие-то фильтры, а applicableFilters. Здесь же только те фильтры, которые нам подходят по логике

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+        List<UserFilter> userFiltersList = userFilters.stream()
+                .filter(userFilter -> userFilter.isApplicable(filters))
+                .toList();
+        List<User> finalUserList = followee.toList();
```
finalUserList  говорит мне о том, что это будет неизменяемый список юзером 
отчасти это правильно, но обычно название поля не должно отражать такой логики. 

если ты имеешь в виду, что здесь лежит результат, давай назовем список filteredUsers

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+
+    public long getFollowersCount(long followeeId) {
```
а почему тут нет транзакции? хехе

InnMish commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+
+    public long getFollowingCount(long followerId) {
```
и тут

InnMish commented on src/main/java/school/faang/user_service/service/userFilters/CityFilter.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Stream<User> apply(Stream<User> userStream, UserFilterDto userFilterDto) {
+        return userStream.filter(filter -> filter.getCity().contains(userFilterDto.getCityPattern()));
```
а точно город через contains хочешь проверять? 
я бы на равенство проверяла

InnMish commented on src/main/java/school/faang/user_service/service/userFilters/CountryFilter.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Stream<User> apply(Stream<User> userStream, UserFilterDto userFilterDto) {
+        return userStream.filter(filter -> filter.getCountry().getTitle().contains(userFilterDto.getCountryPattern()));
```
то же самое, как будто надо нее через contains 
город - есть город

InnMish commented on src/main/java/school/faang/user_service/service/userFilters/EmailFilter.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public Stream<User> apply(Stream<User> userStream, UserFilterDto userFilterDto) {
+        return userStream.filter(filter -> filter.getEmail().contains(userFilterDto.getEmailPattern()));
```
тут тоже не через contains

InnMish commented on src/test/java/school/faang/user_service/service/UserServiceTest.java:
если он тебе не нужен, сноси его совсем, а то empty file вроде остался

InnMish commented on src/main/java/school/faang/user_service/mapper/UserMapper.java:
```diff
@@ -0,0 +1,20 @@
+
+    List<User> toEntities(List<UserDto> dtos);
+
+    List<UserDto> toDTOs(List<User> entities);
```
эти два метода ты в приложении не используешь, зачем они здесь? 
наперед ничего не пишем - высока вероятность написать то, чем никто никогда не будет пользоваться)

InnMish commented on src/main/java/school/faang/user_service/mapper/UserMapper.java:
```diff
@@ -0,0 +1,20 @@
+
+    User toEntity(UserDto dto);
+
+    UserDto toDTO(User entity);
```
toDto 
у тебя же класс называется Dto

InnMish commented on src/test/java/school/faang/user_service/service/SubscriptionServiceTest.java:
```diff
@@ -0,0 +1,156 @@
+    private List<UserFilter> userFilter;
+
+    @Mock
+    private UserFilter nameFilter;
```
фильтры тестируем отдельно 

по тесту на каждый фильтр, потом несколько фильтров вместе

fraifree replied:
Пока не меняю тесты, если по коду вопросов не будет, то уже плотно их переработаю

fraifree replied:
Сделал отдельные тесты для фильтров и миксанул парочку

InnMish commented:
Я вижу, что ты используешь аннотации для контроллера, но при этом вижу, что код через постман ты не тестировал, так как он не  работает. Я запустила и проверила) 

Если ты знаешь, как работает контроллер, весь твой код должен быть протестирован запросами через постман. Иначе я задачу не пропущу

fraifree committed:
fix after review

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,75 @@
+        return subscriptionService.getFollowersCount(userId);
+    }
+
+    @GetMapping(value = "/get/{userId}/following/count")
```
можно value опустить, сразу значение в кавычках)

fraifree replied:
Хорошо

InnMish commented on src/test/java/school/faang/user_service/controller/SubscriptionControllerTest.java:
```diff
@@ -0,0 +1,75 @@
+
+
+    @Test
+    void testUnfollowUser() {
```
по неймингу тестов правило общее: что делаем (test) + что тестируем (UnfollowUser) + что хотим получить (sussesfullyUnfollowed) например
то есть чтобы название отражало то, что происходит в тесте

InnMish replied:
вижу, что в сервисе у тебя норм называются, но тут давай тоже чуть понятнее назовем

fraifree replied:
Учту на будущее)))

fraifree committed:
fix after review
create UserService and UserController for testing request by Post Service module
create test by filters

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    @GetMapping("/get/{userId}/followers")
```
у нас из запроса уже понятно, что get  
не нужно в пути еще раз писать, это не по REST  

по ресту мы в пути описываем именно сущность, с которой работаем. Например, подписка/такой_пользователь/подписчики 

а get post update и подобное не пишем 
вообще глаголы практически не используются

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,74 @@
+        return subscriptionService.getFollowers(userId, filter);
+    }
+
+    @GetMapping("/get/{userId}/following")
```
тут тоже get убираем

InnMish commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,74 @@
+        return subscriptionService.getFollowersCount(userId);
+    }
+
+    @GetMapping("/get/{userId}/following/count")
```
и тут убираем
